### PR TITLE
Adding dynamic MAC reading in test_pfcwd_function.py

### PR DESCRIPTION
**Pull request by rraghav-cisco**
The tests in test_pfcwd_function.py are failing since the reverse path is using the source port's MAC address instead of the dest port's mac address. This PR is to dynamically pass the correct mac address for the reverse path test.
@abdosi : pls review.

### CODE REVIEW

**Comment by rraghav-cisco** [comment]
@abdosi , @neethajohn : The failure in the above checks is coming from bgp_bbr test. This PR has no relation to that failure. Pls resubmit or evaluate this PR accordingly.
thanks,
rraghav

File: tests/pfcwd/test_pfcwd_function.py [hunk 1]
```diff
@@ -494,7 +502,7 @@ def verify_tx_egress(self, action):
         dst_port = "[" + str(self.pfc_wd_test_port_id) + "]"
         if action == "forward" and type(self.pfc_wd_test_port_ids) == list:
             dst_port = "".join(str(self.pfc_wd_test_port_ids)).replace(',', '')
-        ptf_params = {'router_mac': self.router_mac,
+        ptf_params = {'router_mac': self.mac_map[self.pfc_wd_rx_port_id[0]],
```
Comment (abdosi) [inline tests/pfcwd/test_pfcwd_function.py:505]: will this not work `router_mac = self.tx_mac` ?

File: tests/pfcwd/test_pfcwd_function.py [hunk 2]
```diff
@@ -475,12 +476,19 @@ def __init__(self, ptf, router_mac, pfc_params, is_dualtor):
...
             self.vlan_mac = router_mac
-        # Verify traffic before pfc storm
-        self.verify_rx_ingress("forward")
+
+    def update_macs(self, port, mac_address):
+        try:
+            self.mac_map
+        except AttributeError:
+            self.mac_map = {}
+        self.mac_map[port] = mac_address
```
Comment (abdosi) [inline tests/pfcwd/test_pfcwd_function.py:491]: do we need this change ?

**Comment by abdosi** [comment]
@rraghav-cisco  please conclude on if we need this many changes or above change should be good enough ?

**Comment by abdosi** [comment]
@rraghav-cisco can we conclude on this PR

**Comment by rraghav-cisco** [comment]
> @rraghav-cisco can we conclude on this PR

@abdosi, I will test and update the diffs. ETA today.

**Comment by rraghav-cisco** [comment]
> > @rraghav-cisco can we conclude on this PR
> 
> @abdosi, I will test and update the diffs. ETA today.

Sorry for the delay. I have updated the PR.

File: tests/pfcwd/test_pfcwd_function.py [hunk 3]
```diff
@@ -613,8 +615,8 @@ def verify_other_pfc_pg(self):
         else:
             other_pg = self.pfc_queue_index + 1
 
-        ptf_params = {'router_mac': self.router_mac,
-                      'vlan_mac': self.vlan_mac,
+        ptf_params = {'router_mac': self.tx_mac,
```
Comment (abdosi) [inline tests/pfcwd/test_pfcwd_function.py:618]: does not look correct

**Comment by rraghav-cisco** [comment]
@abdosi : Any comments?

File: tests/pfcwd/test_pfcwd_function.py (hunk 3 above)
Comment (abdosi) [reply tests/pfcwd/test_pfcwd_function.py:618]: can you check on this ?

Comment (abdosi) [reply tests/pfcwd/test_pfcwd_function.py:618]: @rraghav-cisco can you check on this

**Review by abdosi — APPROVED** [review]

**Comment by rraghav-cisco** [comment]
> @rraghav-cisco PR conflicts with 202205 branch
Raised: https://github.com/sonic-net/sonic-mgmt/pull/10893

**Comment by rraghav-cisco** [comment]
@mssonicbld : 202205 PR: #10893.

**Comment by rraghav-cisco** [comment]
@mssonicbld : 202305 PR: #11022

**Comment by lipxu** [comment]
Hi, @rraghav-cisco and @abdosi , have you tried to test the PR on dualtor devices, seems all test_pfcwd_function related cases failed on dualtor devices after the PR merged, may you please help to check, thanks a lot.

@lolyu @StormLiangMS for vis

**Comment by rraghav-cisco** [comment]
@lipxu : No , I have not tested on dualtor. what error are you getting ?

**Comment by lipxu** [comment]
> @lipxu : No , I have not tested on dualtor. what error are you getting ?

Thanks for looking at this failure, all cases failed in "self.verify_rx_ingress("forward")" during SendVerifyTraffic.
It should related to the "'vlan_mac'" change, thanks.

  File "/usr/local/lib/python2.7/dist-packages/_pytest/python.py", line 174, in pytest_pyfunc_call
    testfunction(**testargs)
  File "/azp/_work/4/s/tests/pfcwd/test_pfcwd_function.py", line 797, in test_pfcwd_actions
    self.is_dualtor)
  File "/azp/_work/4/s/tests/pfcwd/test_pfcwd_function.py", line 478, in __init__
    self.verify_rx_ingress("forward")
  **File "/azp/_work/4/s/tests/pfcwd/test_pfcwd_function.py", line 539, in verify_rx_ingress**
    log_file=log_file)
  File "/azp/_work/4/s/tests/ptf_runner.py", line 80, in ptf_runner
    result = host.shell(cmd, chdir="/root", module_ignore_errors=module_ignore_errors)

class SendVerifyTraffic():
    """ PTF test """
    def __init__(self, ptf, router_mac, tx_mac, pfc_params, is_dualtor):
        """
        Args:
            ptf(AnsibleHost) : ptf instance
            router_mac(string) : router mac address
            ptf_params(dict) : all PFC test params specific to the DUT port
        """
        self.ptf = ptf
        self.router_mac = router_mac
        self.tx_mac = tx_mac
        self.pfc_queue_index = pfc_params['queue_index']
        self.pfc_wd_test_pkt_count = pfc_params['test_pkt_count']
        self.pfc_wd_rx_port_id = pfc_params['rx_port_id']
        self.pfc_wd_test_port = pfc_params['test_port']
        self.pfc_wd_test_port_id = pfc_params['test_port_id']
        self.pfc_wd_test_port_ids = pfc_params['test_port_ids']
        self.pfc_wd_test_neighbor_addr = pfc_params['test_neighbor_addr']
        self.pfc_wd_rx_neighbor_addr = pfc_params['rx_neighbor_addr']
        self.pfc_wd_test_port_vlan_id = pfc_params['test_port_vlan_id']
        self.pfc_wd_rx_port_vlan_id = pfc_params['rx_port_vlan_id']
        self.port_id_to_type_map = pfc_params['port_id_to_type_map']
        self.port_type = pfc_params['port_type']
        if is_dualtor:
            self.vlan_mac = "00:aa:bb:cc:dd:ee"
        else:
            self.vlan_mac = router_mac
        # Verify traffic before pfc storm
        **self.verify_rx_ingress("forward")**

**Comment by rraghav-cisco** [comment]
@lipxu 
I am surprised. the mac address is modified only for the non-dualtor case. As you can see, the above code sets mac to a hardcoded value in dualtor case. This PR shouldn't affect dualtor.

**Comment by lipxu** [comment]
> @lipxu I am surprised. the mac address is modified only for the non-dualtor case. As you can see, the above code sets mac to a hardcoded value in dualtor case. This PR shouldn't affect dualtor.

The cases in test_pfcwd_function would also run on dualtor devices. This PR modified the "'vlan_mac': self.vlan_mac," to "'vlan_mac': self.tx_mac,", which caused failure on dualtor devices, please help to check and make sure the cases could pass on dualtor devices, thanks.

**Comment by lipxu** [comment]
> > @lipxu I am surprised. the mac address is modified only for the non-dualtor case. As you can see, the above code sets mac to a hardcoded value in dualtor case. This PR shouldn't affect dualtor.
> 
> The cases in test_pfcwd_function would also run on dualtor devices. This PR modified the "'vlan_mac': self.vlan_mac," to "'vlan_mac': self.tx_mac,", which caused failure on dualtor devices, please help to check and make sure the cases could pass on dualtor devices, thanks.

Hi, @rraghav-cisco , Do you plan to fix the issue on dualtor? thanks.

BTW, may you share us the original failure of this issue, is it a special issue of cisco devices? or a special issue of T2? or a common issue on all platforms? thanks.

**Comment by rraghav-cisco** [comment]
> > > @lipxu I am surprised. the mac address is modified only for the non-dualtor case. As you can see, the above code sets mac to a hardcoded value in dualtor case. This PR shouldn't affect dualtor.
> > 
> > 
> > The cases in test_pfcwd_function would also run on dualtor devices. This PR modified the "'vlan_mac': self.vlan_mac," to "'vlan_mac': self.tx_mac,", which caused failure on dualtor devices, please help to check and make sure the cases could pass on dualtor devices, thanks.
> 
> Hi, @rraghav-cisco , Do you plan to fix the issue on dualtor? thanks.
> 
> BTW, may you share us the original failure of this issue, is it a special issue of cisco devices? or a special issue of T2? or a common issue on all platforms? thanks.

@lipxu : I will work on fixing the issue.
The original problem was this: The original code used mac address of DUT(common mac address) as the dst mac of the packets. But in multi-asic case, the mac address of interfaces changes according to the ASIC they belong to. The PR addresses this difference by simply calling the standard pytest duthost function to get the corect mac address of every interface..

**Comment by lipxu** [comment]
> > > > @lipxu I am surprised. the mac address is modified only for the non-dualtor case. As you can see, the above code sets mac to a hardcoded value in dualtor case. This PR shouldn't affect dualtor.
> > > 
> > > 
> > > The cases in test_pfcwd_function would also run on dualtor devices. This PR modified the "'vlan_mac': self.vlan_mac," to "'vlan_mac': self.tx_mac,", which caused failure on dualtor devices, please help to check and make sure the cases could pass on dualtor devices, thanks.
> > 
> > 
> > Hi, @rraghav-cisco , Do you plan to fix the issue on dualtor? thanks.
> > BTW, may you share us the original failure of this issue, is it a special issue of cisco devices? or a special issue of T2? or a common issue on all platforms? thanks.
> 
> @lipxu : I will work on fixing the issue. The original problem was this: The original code used mac address of DUT(common mac address) as the dst mac of the packets. But in multi-asic case, the mac address of interfaces changes according to the ASIC they belong to. The PR addresses this difference by simply calling the standard pytest duthost function to get the corect mac address of every interface..

Thanks @rraghav-cisco , if this is T2 special case, maybe we can isolate the code with previous ones, thanks. Please let us know if any update.

**Comment by rraghav-cisco** [comment]
@lipxu : I have raised a PR to fix the issue: https://github.com/sonic-net/sonic-mgmt/pull/12003

**Pull request merged**